Python/chapter02/2.8 - Loop Detection/miguel_soln_2.8.py

```diff
@@ -183,8 +183,20 @@ def loop_detection_linear_time_const_space(ll: LinkedList):
     while slow_ptr and fast_ptr and fast_ptr.next:
         slow_ptr = slow_ptr.next
         fast_ptr = fast_ptr.next.next
-        if slow_ptr is fast_ptr or slow_ptr is fast_ptr.next:
-            return slow_ptr
+        if slow_ptr is not fast_ptr:
+            continue
+        # if we get here, then there is a cycle.
+        # advance one of fast or slow pointers
```

Comment:
This comment is mostly explaining what the code does, which is not super useful. It also doesn't _quite_ reflect the code, since you say "advance one of fast or slow pointers" but you chose to advance the slow pointer.

Would it be possible to explain _why_ / _how_ just advancing the pointers at the same speed will always make them both align at the beginning of the cycle?

As a hint: try to decompose the position of the fast and slow pointers in terms of the length of the list leading to the cycle, the length of the cycle, and the offset from the start of the cycle. Sprinkle in some modular arithmetic and you should be able to come up with a rough proof of correctness.